crates/sparrow-sources/src/in_memory.rs

```diff
@@ -222,3 +224,180 @@ impl InMemoryBatches {
+        let in_mem = InMemoryBatches::new(true, schema.clone());
+
+        let mut s1 = in_mem.subscribe();
+        let mut s2 = in_mem.subscribe();
+
+        // Add the first batch
+        in_mem.add_batch(batch1.clone()).await.unwrap();
+
+        let b1_s1 = s1.next().await.unwrap().unwrap();
+        assert_eq!(batch1, b1_s1);
+        let b1_s2 = s2.next().await.unwrap().unwrap();
+        assert_eq!(batch1, b1_s2);
+
+        // Add the second batch
+        in_mem.add_batch(batch2.clone()).await.unwrap();
+        let b2_s1 = s1.next().await.unwrap().unwrap();
+        assert_eq!(batch2, b2_s1);
+        let b2_s2 = s2.next().await.unwrap().unwrap();
+        assert_eq!(batch2, b2_s2);
+    }
+
+    #[tokio::test]
+    async fn test_subscribe_to_multiple_batches() {
+        // Sends multiple batches before reading
+
+        let time: ArrayRef = Arc::new(TimestampNanosecondArray::from(vec![0, 1, 2]));
+        let subsort: ArrayRef = Arc::new(UInt64Array::from(vec![0, 1, 2]));
+        let key: ArrayRef = Arc::new(UInt64Array::from(vec![0, 0, 0]));
+        let a: ArrayRef = Arc::new(Int32Array::from(vec![0, 1, 2]));
+        let b: ArrayRef = Arc::new(UInt64Array::from(vec![None, Some(1), Some(8)]));
+        let schema = Schema::new(vec![
+            Field::new("_time", TimestampNanosecondType::DATA_TYPE, false),
+            Field::new("_subsort", DataType::UInt64, false),
+            Field::new("_key_hash", DataType::UInt64, false),
+            Field::new("a", a.data_type().clone(), false),
+            Field::new("b", b.data_type().clone(), true),
+        ]);
+        let schema = Arc::new(schema);
+        let batch1 = RecordBatch::try_new(
+            schema.clone(),
+            vec![time, subsort.clone(), key.clone(), a, b],
+        )
+        .unwrap();
+
+        let time2: ArrayRef = Arc::new(TimestampNanosecondArray::from(vec![3, 4, 5]));
+        let a2: ArrayRef = Arc::new(Int32Array::from(vec![3, 4, 5]));
+        let b2: ArrayRef = Arc::new(UInt64Array::from(vec![Some(10), None, None]));
+        // Can reuse the subsort/key arrays
+        let batch2 =
+            RecordBatch::try_new(schema.clone(), vec![time2, subsort, key, a2, b2]).unwrap();
+
+        let in_mem = InMemoryBatches::new(true, schema.clone());
+
+        let mut s1 = in_mem.subscribe();
+        let mut s2 = in_mem.subscribe();
+
+        // Add both batches before reading
+        in_mem.add_batch(batch1.clone()).await.unwrap();
+        in_mem.add_batch(batch2.clone()).await.unwrap();
+
+        let b1_s1 = s1.next().await.unwrap().unwrap();
+        assert_eq!(batch1, b1_s1);
+        let b1_s2 = s2.next().await.unwrap().unwrap();
+        assert_eq!(batch1, b1_s2);
+
+        let b2_s1 = s1.next().await.unwrap().unwrap();
+        assert_eq!(batch2, b2_s1);
+        let b2_s2 = s2.next().await.unwrap().unwrap();
+        assert_eq!(batch2, b2_s2);
+    }
+
+    #[tokio::test]
+    async fn test_late_subscription_receives_merged_batch() {
```

Comment:
Should there be one for the non-subscription case (eg., just a read)? What about the case where we don't have retention?